ring_theory/determinants.lean

```diff
@@ -0,0 +1,213 @@
+  {β} [comm_monoid β] {f : α → β} {s : finset α} (x : α) :
+  s.prod f =
+  (s.map ⟨λ z, z * x⁻¹, λ _ _, mul_right_cancel⟩).prod (λ z, f (z * x)) :=
+finset.prod_bij (λ z _, z * x⁻¹) (λ _ _, by simpa) (λ _ _, by simp)
+  (λ _ _ _ _, mul_right_cancel) (λ b H,
+    ⟨b * x, by revert H; simp [eq_comm] {contextual:=tt}, by simp⟩)
+
+lemma finset.sum_mul_right {α} [group α]
+  {β} [add_comm_monoid β] {f : α → β} {s : finset α} (x : α) :
+  s.sum f =
+  (s.map ⟨λ z, z * x⁻¹, λ _ _, mul_right_cancel⟩).sum (λ z, f (z * x)) :=
+finset.sum_bij (λ z _, z * x⁻¹) (λ _ _, by simpa) (λ _ _, by simp)
+  (λ _ _ _ _, mul_right_cancel) (λ b H,
+    ⟨b * x, by revert H; simp [eq_comm] {contextual:=tt}, by simp⟩)
+
+lemma finset.univ_sum_mul_right {α} [group α] [fintype α]
+  {β} [add_comm_monoid β] {f : α → β} (x : α) :
+  finset.univ.sum f =
+  finset.univ.sum (λ z, f (z * x)) :=
+finset.sum_bij (λ z _, z * x⁻¹) (λ _ _, finset.mem_univ _) (λ _ _, by simp)
+  (λ _ _ _ _, mul_right_cancel) (λ b H,
+    ⟨b * x, by revert H; simp [eq_comm] {contextual:=tt}, by simp⟩)
+
+lemma finset.prod_perm {α} [fintype α] [decidable_eq α]
+  {β} [comm_monoid β] {f : α → β} (σ : equiv.perm α) :
+  (finset.univ : finset α).prod f
+  = finset.univ.prod (λ z, f (σ z)) :=
+eq.symm $ finset.prod_bij (λ z _, σ z) (λ _ _, finset.mem_univ _) (λ _ _, rfl)
+  (λ _ _ _ _ H, σ.bijective.1 H) (λ b _, ⟨σ⁻¹ b, finset.mem_univ _, by simp⟩)
+
+instance {α} (H : α → Prop) : subsingleton (decidable_pred H) :=
+by apply_instance
+
+@[simp] lemma finset.filter_true {α} (s : finset α) [h : decidable_pred (λ _, true)] :
+  @finset.filter α (λ _, true) h s = s :=
+by ext; simp
+
+namespace matrix
+open equiv
+variables {n : Type u} [fintype n] [decidable_eq n] {R : Type v} [comm_ring R]
+
+instance : group (equiv.perm n) := by apply_instance
+
+@[simp] lemma equiv.swap_mul_self (i j : n) : equiv.swap i j * equiv.swap i j = 1 :=
+equiv.swap_swap i j
+
+@[simp] lemma equiv.swap_swap_apply (i j k : n) : equiv.swap i j (equiv.swap i j k) = k :=
+equiv.swap_core_swap_core k i j
+
+instance : decidable_pred (function.bijective : (n → n) → Prop) :=
+λ _, by unfold function.bijective; apply_instance
+
+instance bij_fintype : fintype {f : n → n // function.bijective f} := 
+set_fintype _
+
+@[extensionality] theorem equiv.perm.ext (σ τ : equiv.perm n)
+  (H : ∀ i, σ i = τ i) : σ = τ :=
+equiv.ext _ _ H
+
+@[simp] lemma equiv.perm.sign_mul (σ τ : equiv.perm n) :
+  (σ * τ).sign = σ.sign * τ.sign :=
+is_group_hom.mul _ _ _
+
+@[simp] lemma equiv.perm.sign_one :
+  equiv.perm.sign (1 : equiv.perm n) = 1 :=
+is_group_hom.one _
+
+@[simp] lemma equiv.perm.sign_refl :
+  equiv.perm.sign (equiv.refl n) = 1 :=
+is_group_hom.one equiv.perm.sign
+
+@[simp] lemma equiv.perm.sign_swap' {i j : n} :
+  (equiv.swap i j).sign = if i = j then 1 else -1 :=
+if H : i = j then by simp [H, equiv.swap_self] else
+by simp [equiv.perm.sign_swap H, H]
+
+def e (σ : equiv.perm n) : R := ((σ.sign : ℤ) : R)
```

Comment:
You want to have `e` in the `matrix` namespace? It isn't even related to matices.